Fix #291: Add Retry-After http header in 509 response

DavidKarlas opened the pull request:
Let me know how ugly solution it is from 0 to 10 to do `if (e.code() == 509)` and then `dynamic_cast`. It shows that my language is C# :P

tomhughes commented:
It theory it's safe to use `static_cast` there because we know from the code check that the object should be of that type.

Now you can argue that `dynamic_cast` verifies that but as this is a reference it does so by throwing an exception though I guess that just means this request will abort - the alternative would be to do a pointer case instead and then just not add the header if you get null.

DavidKarlas commented:
Did dynamic_cast of pointer, and added if, to not crash, thank you!
Fixed unit tests, and reverted change of 404 return type from text/plain to text/html because tests started failing... Hence I think it should be separate PR...

mmd-osm commented on src/options.cpp:
```diff
@@ -147,17 +147,17 @@ void global_settings_via_options::set_oauth_10_support(const po::variables_map &
-    if (parsed_bytes_per_sec < 0)
-      throw std::invalid_argument("ratelimit must be a positive number");
+    if (parsed_bytes_per_sec <= 0)
+      throw std::invalid_argument("ratelimit must be greater than zero");
```
Good catch. I thought about some division by zero issue in rate_limiter.cpp, but later forgot about it again.

mmd-osm commented:
Thanks! Looks good to me, let's give it a try and merge!
